bitcoin/src/address.rs

```diff
@@ -643,9 +505,193 @@ impl NetworkValidation for NetworkUnchecked {
 /// representation of an address without the network validation tag, which is used to ensure that
 /// addresses are used only on the appropriate network.
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-struct AddressInner {
-    payload: Payload,
-    network: Network,
+enum AddressInner {
+    P2pkh {
+        hash: PubkeyHash,
+        prefix: LegacyP2pkhPrefix,
+    },
+    P2sh {
+        hash: ScriptHash,
+        prefix: LegacyP2shPrefix,
+    },
+    Segwit {
+        version: WitnessVersion,
+        /// Guaranteed to be between 2 and 40 bytes.
+        program: PushBytesBuf,
```

Comment:
Can `version` and `program` be combined into `WitnessProgram` here?

Reply:
Would make review substaincially easier as we don't need to care about checking 2 and 40 bytes invariant.

Reply:
It'd affect the rest of the code and I'm not sure if it helps. I'm just going AFK so can't look into it.

Reply:
There is already a `WitnessProgram` structure that has the guarantee that bytes are 2 and 40 inclusive. Seems less error-prone to me. I could easily see someone manually assigning `AddressInner` here without checking the lens.

Reply:
Hey @sanket1729 if you get time could you take a look at #1846, just the concept, thanks (its related to this thread).